File: user-service/src/main/java/rs/edu/raf/mapper/ClientMapper.java
package rs.edu.raf.mapper;

import lombok.AllArgsConstructor;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.stereotype.Component;
import rs.edu.raf.dto.ChangePasswordWithCodeDTO;
import rs.edu.raf.dto.ClientDTO;
import rs.edu.raf.dto.ClientCreateDTO;
import rs.edu.raf.dto.ClientRegisterDTO;
import rs.edu.raf.exceptions.UserNotFoundException;
import rs.edu.raf.model.user.Client;
import rs.edu.raf.repository.ClientRepository;
import rs.edu.raf.repository.RoleRepository;

import java.util.Optional;

@Component
@AllArgsConstructor
public class ClientMapper {
    private ClientRepository clientRepository;
    private BCryptPasswordEncoder bCryptPasswordEncoder;
    private RoleRepository roleRepository;

    public Client clientCreateDTOtoClient(ClientCreateDTO clientCreateDTO) {
        Client client = new Client();

        client.setFirstName(clientCreateDTO.firstName());
        client.setLastName(clientCreateDTO.lastName());
        client.setJMBG(clientCreateDTO.JMBG());
        client.setDateOfBirth(clientCreateDTO.dateOfBirth());
        client.setGender(clientCreateDTO.gender());
        client.setEmail(clientCreateDTO.email());
        client.setPhoneNumber(clientCreateDTO.phoneNumber());
        client.setAddress(clientCreateDTO.address());
        client.setActive(clientCreateDTO.active());
        client.getRoles().add(roleRepository.findByName("ROLE_CLIENT").get());

        return client;
    }

    public Client clientRegisterDTOtoClient(ClientRegisterDTO clientRegisterDTO) {

        Optional<Client> client = clientRepository.findByEmailAndActiveIsTrue(clientRegisterDTO.email());

        if (client.isPresent()){
            client.get().setPassword(bCryptPasswordEncoder.encode(clientRegisterDTO.password()));
            return client.get();
        }

        throw new UserNotFoundException(clientRegisterDTO.email());
    }

    public ClientDTO clientToClientDTO(Client client) {
        return new ClientDTO(client.getId(),
            client.getFirstName(),
            client.getLastName(),
            client.getJMBG(),
            client.getDateOfBirth(),
            client.getGender(),
            client.getEmail(),
            client.getPhoneNumber(),
            client.getAddress(),
            client.getAccountNumbers()
        );
    }

    public Client changePasswordDTOtoClient(ChangePasswordWithCodeDTO changePasswordWithCodeDTO) {
        Optional<Client> client = clientRepository.findByEmailAndActiveIsTrue(changePasswordWithCodeDTO.email());
        if(client.isPresent()) {
            client.get().setPassword(bCryptPasswordEncoder.encode(changePasswordWithCodeDTO.password()));
            return client.get();
        }
        throw new UserNotFoundException(changePasswordWithCodeDTO.email());
    }

}

File: user-service-kotlin/src/main/kotlin/org/example/userservicekotlin/mapper/ClientMapper.kt
package org.example.userservicekotlin.mapper

import org.example.userservicekotlin.dto.ClientCreateDTO
import org.example.userservicekotlin.dto.ClientDTO
import org.example.userservicekotlin.dto.ClientRegisterDTO
import org.example.userservicekotlin.dto.ChangePasswordWithCodeDTO
import org.example.userservicekotlin.model.user.Client
import org.example.userservicekotlin.repository.ClientRepository
import org.example.userservicekotlin.repository.RoleRepository
import org.example.userservicekotlin.exception.UserNotFoundException
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import org.springframework.stereotype.Component
import java.util.*

@Component
class ClientMapper(
    val clientRepository: ClientRepository,
    val roleRepository: RoleRepository,
    val bCryptPasswordEncoder: BCryptPasswordEncoder
) {
    fun clientCreateDTOtoClient(clientCreateDTO: ClientCreateDTO): Client {
        return Client(
            null,
            clientCreateDTO.phoneNumber,
            clientCreateDTO.address,
            clientCreateDTO.firstName,
            clientCreateDTO.lastName,
            clientCreateDTO.JMBG,
            clientCreateDTO.dateOfBirth,
            clientCreateDTO.gender,
            clientCreateDTO.email,
            null,
            clientCreateDTO.active,
            mutableListOf(roleRepository.findByName("ROLE_CLIENT").get()),
            null
        )
    }

    fun clientRegisterDTOtoClient(clientRegisterDTO: ClientRegisterDTO): Client {
        val client: Optional<Client> = clientRepository.findByEmailAndActiveIsTrue(clientRegisterDTO.email)

        if (client.isPresent) {
            client.get().userPassword = bCryptPasswordEncoder.encode(clientRegisterDTO.password)
            return client.get()
        }

        throw UserNotFoundException(clientRegisterDTO.email)
    }

    fun clientToClientDTO(client: Client): ClientDTO {
        return ClientDTO(
            client.id,
            client.firstName,
            client.lastName,
            client.JMBG,
            client.dateOfBirth,
            client.gender,
            client.email,
            client.phoneNumber,
            client.address,
            client.accountNumbers
        )
    }

    fun changePasswordDTOtoClient(changePasswordWithCodeDTO: ChangePasswordWithCodeDTO): Client {
        val client: Optional<Client> = clientRepository.findByEmailAndActiveIsTrue(changePasswordWithCodeDTO.email)
        if (client.isPresent) {
            client.get().userPassword = bCryptPasswordEncoder.encode(changePasswordWithCodeDTO.password)
            return client.get()
        }
        throw UserNotFoundException(changePasswordWithCodeDTO.email)
    }
}